pkg/vsphere/toolbox/vix_command.go

```diff
@@ -0,0 +1,342 @@
+
+	header struct {
+		StartMinimized    uint8
+		ProgramPathLength uint32
+		ArgumentsLength   uint32
+		WorkingDirLength  uint32
+		NumEnvVars        uint32
+		EnvVarLength      uint32
+	}
+
+	ProgramPath string
+	Arguments   string
+	WorkingDir  string
+	EnvVars     []string
+}
+
+type VixCommandHandler func(string, VixCommandRequestHeader, []byte) ([]byte, error)
+
+type VixRelayedCommandHandler struct {
+	Out Channel
+
+	ProcessStartCommand func(*VixMsgStartProgramRequest) (int, error)
+
+	handlers map[uint32]VixCommandHandler
+}
+
+func RegisterVixRelayedCommandHandler(service *Service) *VixRelayedCommandHandler {
+	handler := &VixRelayedCommandHandler{
+		Out:      service.out,
+		handlers: make(map[uint32]VixCommandHandler),
+	}
+
+	service.RegisterHandler("Vix_1_Relayed_Command", handler.Dispatch)
+
+	handler.RegisterHandler(vixCommandGetToolsState, handler.GetToolsState)
+
+	handler.RegisterHandler(vixCommandStartProgram, handler.StartCommand)
+
+	handler.ProcessStartCommand = handler.ExecCommandStart
+
+	return handler
+}
+
+func (c *VixRelayedCommandHandler) Dispatch(data []byte) ([]byte, error) {
+	// See ToolsDaemonTcloGetQuotedString
+	if data[0] == '"' {
+		data = data[1:]
+	}
+
+	var name string
+
+	ix := bytes.IndexByte(data, '"')
+	if ix > 0 {
+		name = string(data[:ix])
+		data = data[ix+1:]
+	}
+	// skip the NULL
+	if data[0] == 0 {
+		data = data[1:]
+	}
+
+	if Trace {
+		fmt.Fprintf(os.Stderr, "vix dispatch '%s'...\n%s\n", name, hex.Dump(data))
+	}
+
+	var header VixCommandRequestHeader
+	buf := bytes.NewBuffer(data)
+	err := binary.Read(buf, binary.LittleEndian, &header)
+	if err != nil {
+		return nil, err
+	}
+
+	if header.Magic != vixCommandMagicWord {
+		return nil, errors.New("VIX_E_INVALID_MESSAGE_HEADER")
+	}
+
+	handler, ok := c.handlers[header.OpCode]
+	if !ok {
+		return nil, errors.New("VIX_E_UNRECOGNIZED_COMMAND_IN_GUEST")
+	}
+
+	creds := buf.Bytes()[header.BodyLength:]
+	// TODO: ignoring credentials for now
+	_ = c.authenticate(header, creds)
+
+	// All Foundry tools commands return results that start with a foundry error
+	// and a guest-OS-specific error (e.g. errno)
+	var rc, errno int
+
+	response, err := handler(name, header, buf.Bytes())
+	if err != nil {
+		// TODO: support the other 10 million VIX_E_* errors
+		rc = 1 // VIX_E_FAIL
+		// TODO: inspect err for system error, setting errno
+	}
+
+	return append([]byte(fmt.Sprintf("%d %d ", rc, errno)), response...), nil
+}
+
+func (c *VixRelayedCommandHandler) RegisterHandler(op uint32, handler VixCommandHandler) {
+	c.handlers[op] = handler
+}
+
+func (c *VixRelayedCommandHandler) GetToolsState(_ string, _ VixCommandRequestHeader, _ []byte) ([]byte, error) {
+	hostname, _ := os.Hostname()
+	osname := fmt.Sprintf("%s-%s", runtime.GOOS, runtime.GOARCH)
+
+	// Note that vmtoolsd sends back 40 or so of these properties, sticking with the minimal set for now.
+	props := VixPropertyList{
+		NewStringProperty(VixPropertyGuestOsVersion, osname),
```

Comment:
You mean the values such as "VMware Tools (Go)"?  In any case, I agree, I'll add some comments/references to where we derive these from open-vm-tools.